File: sdk/python/atb/data_export_gate.py
"""Data export profile workflow helper."""

from __future__ import annotations

import time
from collections.abc import Callable, Mapping
from dataclasses import dataclass
from typing import Any, Literal, TypeVar

from atb.bundle import Bundle
from atb.exceptions import ATBError
from atb.identity_evidence import IdentityEvidence, identity_evidence_payload
from atb.workflow_common import (
    WorkflowContext,
    actor_id_hash,
    canonical_digest,
    new_action_id,
    new_approval_id,
    value_digest,
)

T = TypeVar("T")


class DataExportDeniedError(ATBError):
    """Raised when a data export is denied in enforce mode."""


@dataclass(frozen=True)
class DataExportInput:
    action_type: str
    target_resource_id: str
    intended_effect: str
    action_parameters: Mapping[str, Any]
    subject_id: str | None = None
    action_id: str | None = None
    request_id: str | None = None


@dataclass(frozen=True)
class DataExportApproval:
    approver_id_hash: str
    approval_outcome: Literal["approved", "denied"] = "approved"
    justification_digest: str | None = None
    approval_id: str | None = None
    identity_evidence: IdentityEvidence | None = None


class DataExportGate:
    def __init__(
        self,
        bundle: Bundle | None = None,
        *,
        mode: Literal["log_only", "enforce"] = "log_only",
        policy: Callable[[DataExportInput], Mapping[str, Any]] | None = None,
        record_approval: bool | Callable[[DataExportInput], DataExportApproval] = True,
        auto_save: bool = False,
        save_path: str | None = None,
        actor_id: str | None = None,
        org_id: str | None = None,
        workspace_id: str | None = None,
    ) -> None:
        if mode not in {"log_only", "enforce"}:
            raise ValueError("mode must be one of: log_only, enforce")
        self.ctx = WorkflowContext(
            bundle,
            auto_save=auto_save,
            save_path=save_path,
            actor_id=actor_id,
            org_id=org_id,
            workspace_id=workspace_id,
        )
        self.mode = mode
        self.policy = policy or (lambda _: {"decision": "allow", "reason_codes": []})
        self.record_approval = record_approval
        self.actor_id = actor_id

    @property
    def bundle(self) -> Bundle:
        return self.ctx.bundle

    def run(self, export_action: DataExportInput, fn: Callable[[], T]) -> T:
        self.ctx.bootstrap_request("data_export", export_action.request_id)
        action_id = new_action_id(export_action.action_id)

        self.ctx.emit(
            "data.export.precommit",
            {
                "action_id": action_id,
                "action_type": export_action.action_type,
                "action_parameters_digest": canonical_digest(
                    dict(export_action.action_parameters)
                ),
                "target_resource_id": export_action.target_resource_id,
                "intended_effect": export_action.intended_effect,
            },
        )

        decision = self.policy(export_action)
        self.ctx.emit(
            "ai.policy.decision",
            self.ctx.policy_payload(action_id, decision, export_action.subject_id),
        )
        if decision.get("decision") == "deny" and self.mode == "enforce":
            raise DataExportDeniedError(
                f"data export denied by policy for action_id {action_id}"
            )

        # Resolve the approval record before executing the export: a raising
        # record_approval callback (or malformed identity evidence) rejects the
        # action up front instead of masking the export outcome after fn() ran.
        approval_payload = self._build_approval_payload(action_id, export_action)

        started = time.perf_counter()
        try:
            result = fn()
            self.ctx.emit(
                "data.export.executed",
                {
                    "action_id": action_id,
                    "execution_outcome": "success",
                    "tool_receipt_digest": value_digest(result),
                    "execution_duration_ms": max(
                        0, int((time.perf_counter() - started) * 1000)
                    ),
                },
            )
            self._emit_approval_payload(approval_payload)
            return result
        except Exception as exc:
            # A data export that raised did not complete: record the forensic
            # data.export.error event, not a success-shaped executed record.
            self.ctx.emit(
                "data.export.error",
                {
                    "action_id": action_id,
                    "error_class": "exception",
                    "error_detail_digest": value_digest(exc),
                },
            )
            self._emit_approval_payload(approval_payload)
            raise

    def _build_approval_payload(
        self, action_id: str, export_action: DataExportInput | None = None
    ) -> dict[str, Any] | None:
        if not self.record_approval:
            return None
        approval = (
            self.record_approval(export_action)
            if callable(self.record_approval) and export_action is not None
            else DataExportApproval(
                approver_id_hash=actor_id_hash(self.actor_id),
                justification_digest=canonical_digest({"reason": "export approved"}),
            )
        )
        payload: dict[str, Any] = {
            "approval_id": new_approval_id(approval.approval_id),
            "approver_id_hash": approval.approver_id_hash,
            "approval_outcome": approval.approval_outcome,
            "justification_digest": approval.justification_digest
            or canonical_digest({"reason": "export approved"}),
            "action_id": action_id,
        }
        identity_evidence = identity_evidence_payload(approval.identity_evidence)
        if identity_evidence is not None:
            payload["identity_evidence"] = identity_evidence
        return payload

    def _emit_approval_payload(self, payload: dict[str, Any] | None) -> None:
        if payload is not None:
            self.ctx.emit("ai.human.approval", payload)

File: sdk/typescript/src/data-export-gate.ts
import {
  WorkflowContext,
  type WorkflowContextOptions,
  type WorkflowPolicyDecision,
  actorIdHash,
  canonicalDigest,
  newActionId,
  newApprovalId,
  valueDigest,
} from "./workflow-common.js";
import { identityEvidencePayload, type IdentityEvidence } from "./identity-evidence.js";

export type DataExportGateMode = "log_only" | "enforce";

/** Input metadata for a gated data export. */
export interface DataExportInput {
  actionType: string;
  targetResourceId: string;
  intendedEffect: string;
  actionParameters: Record<string, unknown>;
  subjectId?: string;
  actionId?: string;
  requestId?: string;
}

/** Human approval metadata for export workflows. */
export interface DataExportApproval {
  approverIdHash: string;
  approvalOutcome?: "approved" | "denied";
  justificationDigest?: string;
  approvalId?: string;
  /** Advanced: digest-only caller-provided reviewer identity evidence. */
  identityEvidence?: IdentityEvidence;
}

export interface DataExportGateOptions extends WorkflowContextOptions {
  mode?: DataExportGateMode;
  policy?: (
    exportAction: DataExportInput
  ) => WorkflowPolicyDecision | Promise<WorkflowPolicyDecision>;
  /** When true, emit ai.human.approval after execution (required by profile when executed). */
  recordApproval?: boolean | ((exportAction: DataExportInput) => DataExportApproval);
}

export class DataExportDeniedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DataExportDeniedError";
  }
}

/** Records data export profile events around an export operation. */
export class DataExportGate {
  readonly ctx: WorkflowContext;
  private readonly mode: DataExportGateMode;
  private readonly actorId?: string;
  private readonly policy: (
    exportAction: DataExportInput
  ) => WorkflowPolicyDecision | Promise<WorkflowPolicyDecision>;
  private readonly recordApproval:
    | boolean
    | ((exportAction: DataExportInput) => DataExportApproval);

  constructor(options: DataExportGateOptions = {}) {
    this.ctx = new WorkflowContext(options);
    this.actorId = options.actorId;
    this.mode = options.mode ?? "log_only";
    this.policy =
      options.policy ??
      (() => ({
        decision: "allow",
        policyId: "local.data_export",
        policyVersion: "v1",
        reasonCodes: [],
      }));
    this.recordApproval = options.recordApproval ?? true;
    if (!["log_only", "enforce"].includes(this.mode)) {
      throw new Error("mode must be one of: log_only, enforce");
    }
  }

  get bundle() {
    return this.ctx.bundle;
  }

  async run<T>(
    exportAction: DataExportInput,
    fn: () => T | Promise<T>
  ): Promise<T> {
    this.ctx.bootstrapRequest("data_export", exportAction.requestId);
    const actionId = newActionId(exportAction.actionId);

    this.ctx.emit("data.export.precommit", {
      action_id: actionId,
      action_type: exportAction.actionType,
      action_parameters_digest: canonicalDigest(exportAction.actionParameters),
      target_resource_id: exportAction.targetResourceId,
      intended_effect: exportAction.intendedEffect,
    });

    const decision = await this.policy(exportAction);
    this.ctx.emit(
      "ai.policy.decision",
      this.ctx.policyPayload(actionId, decision, exportAction.subjectId)
    );
    if (decision.decision === "deny" && this.mode === "enforce") {
      throw new DataExportDeniedError(
        `data export denied by policy for action_id ${actionId}`
      );
    }

    const startedAt = Date.now();
    try {
      const result = await fn();
      this.ctx.emit("data.export.executed", {
        action_id: actionId,
        execution_outcome: "success",
        tool_receipt_digest: valueDigest(result),
        execution_duration_ms: Math.max(0, Date.now() - startedAt),
      });
      this.maybeRecordApproval(exportAction, actionId);
      return result;
    } catch (error) {
      // A data export that threw did not complete: record the forensic
      // data.export.error event, not a success-shaped executed record.
      this.ctx.emit("data.export.error", {
        action_id: actionId,
        error_class: "exception",
        error_detail_digest: valueDigest(error),
      });
      this.maybeRecordApproval(exportAction, actionId);
      throw error;
    }
  }

  private maybeRecordApproval(
    exportAction: DataExportInput,
    actionId: string
  ): void {
    if (this.recordApproval === false) {
      return;
    }
    const approval =
      typeof this.recordApproval === "function"
        ? this.recordApproval(exportAction)
        : {
            approverIdHash: actorIdHash(this.actorId),
            approvalOutcome: "approved" as const,
            justificationDigest: canonicalDigest({ reason: "export approved" }),
          };
    const payload: Record<string, unknown> = {
      approval_id: newApprovalId(approval.approvalId),
      approver_id_hash: approval.approverIdHash,
      approval_outcome: approval.approvalOutcome ?? "approved",
      justification_digest:
        approval.justificationDigest ?? canonicalDigest({ reason: "export approved" }),
      action_id: actionId,
    };
    const identityEvidence = identityEvidencePayload(approval.identityEvidence);
    if (identityEvidence) {
      payload.identity_evidence = identityEvidence;
    }
    this.ctx.emit("ai.human.approval", payload);
  }
}
